plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/LayerConstraintPreprocessor.java

```diff
@@ -0,0 +1,248 @@
+        case FIRST_SEPARATE:
+        case LAST_SEPARATE:
+            return true;
+            
+        default:
+            return false;
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Node Hiding
+    
+    /**
+     * Disconnects the node from its adjacent nodes.
+     */
+    private void hide(final LNode lNode) {
+        ensureNoInacceptableEdges(lNode);
+        
+        for (LEdge lEdge : lNode.getConnectedEdges()) {
+            hide(lNode, lEdge);
+        }
+    }
+
+    /**
+     * Hides the edge of the given node. The node is the one that's currently being hidden.
+     */
+    private void hide(final LNode lNode, final LEdge lEdge) {
+        boolean outgoing = lEdge.getSource().getNode() == lNode;
+        LPort oppositePort = outgoing
+                ? lEdge.getTarget()
+                : lEdge.getSource();
+        
+        // Hide the edge and remember its original target or source
+        if (outgoing) {
+            lEdge.setTarget(null);
+        } else {
+            lEdge.setSource(null);
+        }
+        
+        lEdge.setProperty(InternalProperties.ORIGINAL_OPPOSITE_PORT, oppositePort);
+        
+        // We may want to control where the other node will end up in the layering...
+        updateOppositeNodeLayerConstraints(lNode, oppositePort.getNode());
+    }
+    
+    /**
+     * Used to keep track of which kinds of hidden nodes a given node had connections to.
+     */
+    private static enum HiddenNodeConnections {
+        /** No connections to hidden nodes found yet. */
+        NONE,
+        /** Only connections to {@link LayerConstraint#FIRST_SEPARATE} nodes found so far. */
+        FIRST_SEPARATE,
+        /** Only connections to {@link LayerConstraint#LAST_SEPARATE} nodes found so far. */
+        LAST_SEPARATE,
+        /** Connections to both kinds of hidden nodes found. */
+        BOTH;
+        
+        private HiddenNodeConnections combine(final LayerConstraint layerConstraint) {
+            switch (this) {
+            case NONE:
+                return layerConstraint == LayerConstraint.FIRST_SEPARATE
+                    ? FIRST_SEPARATE
+                    : LAST_SEPARATE;
+            
+            case FIRST_SEPARATE:
+                return layerConstraint == LayerConstraint.FIRST_SEPARATE
+                    ? FIRST_SEPARATE
+                    : BOTH;
+            
+            case LAST_SEPARATE:
+                return layerConstraint == LayerConstraint.FIRST_SEPARATE
+                    ? BOTH
+                    : LAST_SEPARATE;
+            
+            default:
+                return BOTH;
+            }
+        }
+    }
+    
+    /** We'll use this property to keep track of each node's connections to hidden nodes. */
+    private static final IProperty<HiddenNodeConnections> HIDDEN_NODE_CONNECTIONS = new Property<>(
```

Comment:
Why define it here instead of `InternalProperties`?

Reply:
Yeah, it's only used here. I could have used a map to store these information, but I felt that would have cluttered the code more than using this property.